crud/src/main/java/com/redhat/lightblue/assoc/ep/ExecutionPlan.java

```diff
@@ -225,11 +225,13 @@ public ExecutionPlan(QueryExpression requestQuery,
                         if (requestSort != null) {
                             last = new Source<>(new SortResults(block, last, requestSort));
                         }
-                        if (from != null) {
-                            last = new Source<>(new Skip(block, from.intValue(), last));
-                        }
-                        if (to != null) {
-                            last = new Source<>(new Limit(block, to.intValue() - from.intValue() + 1, last));
+                        if(!needsFinalFiltering) {
```

Comment:
Filtering will not always be there. Associations are evaluated in either one or two phases, an optional search phase, and the retrieval phase. If it is  a trivial association, there is no search phase, and everything is done in retrieval phase. But if it is not a trivial association, we first search, find the documents that match the query, and then retrieve those. Depending on the queries, it is possible that not all queries are included in the search phase, meaning it finds more docs than necessary. We add additional filtering only if all association queries cannot be evaluated at the search phase. So, duplication is necessary here. If we can limit during search, we do that. If we can't limit during search, we add a filter, and limit there.